spec/java/util/zip/CRC32.automaton.lsl

```diff
@@ -0,0 +1,202 @@
+libsl "1.1.0";
+
+library std
+    version "11"
+    language "Java"
+    url "https://github.com/openjdk/jdk11/blob/master/src/java.base/share/classes/java/util/zip/CRC32.java";
+
+// imports
+
+import java/nio/ByteBuffer;
+import java/util/zip/Checksum;
+import sun/nio/ch/DirectBuffer;
+import java/util/zip/CRC32;
+
+// automata
+
+automaton CRC32Automaton
+(
+)
+: LSLCRC32
+{
+    // states and shifts
+
+    initstate Allocated;
+    state Initialized;
+
+    shift Allocated -> Initialized by [
+        // constructors
+        CRC32,
+    ];
+
+    shift Initialized -> self by [
+        // instance methods
+        getValue,
+        reset,
+        update (CRC32, ByteBuffer),
+        update (CRC32, array<byte>),
+        update (CRC32, array<byte>, int, int),
+        update (CRC32, int),
+    ];
+
+    // internal variables
+
+    var crc: int = 0;
+
+
+    // utilities
+
+    proc _updateCheck(b: array<byte>, off: int, len: int): void
+    {
+        if (b == null)
+            action THROW_NEW("java.lang.NullPointerException", []);
+
+        val b_size: int = action ARRAY_SIZE(b);
```

Comment:
`b_size` is always positive.
*hint on `ASSUME`*

Reply:
By using `ASSUME` we can prune unreasonable execution paths from all possible.
`b_size` is an `int`. That means it *can* be anything between `-2147483648` to `2147483647` including `0`.
We (as human programmers) know that array sizes cannot be negative. It is an additional *external* information.

By using `ASSUME` we can simulate unsigned types.